src/main/kotlin/com/refinedmods/refinedstorage/apiimpl/network/NetworkNodeGraph.kt

```diff
@@ -0,0 +1,140 @@
+package com.refinedmods.refinedstorage.apiimpl.network
+
+import com.google.common.collect.Sets
+import com.refinedmods.refinedstorage.api.network.INetwork
+import com.refinedmods.refinedstorage.api.network.INetworkNodeGraph
+import com.refinedmods.refinedstorage.api.network.INetworkNodeGraphListener
+import com.refinedmods.refinedstorage.api.network.INetworkNodeVisitor
+import com.refinedmods.refinedstorage.api.network.node.INetworkNode
+import com.refinedmods.refinedstorage.api.util.Action
+import com.refinedmods.refinedstorage.util.NetworkUtils
+import net.minecraft.block.Block
+import net.minecraft.block.entity.BlockEntity
+import net.minecraft.util.math.BlockPos
+import net.minecraft.util.math.Direction
+import net.minecraft.world.World
+import java.util.*
+import java.util.function.Consumer
+import kotlin.collections.HashSet
+
+class NetworkNodeGraph(
+        private val network: INetwork
+) : INetworkNodeGraph {
+    private var nodes: MutableSet<INetworkNode> = Sets.newConcurrentHashSet()
+    private val listeners: MutableList<INetworkNodeGraphListener> = LinkedList()
+    private val actions: MutableSet<Consumer<INetwork>> = HashSet()
+    private var invalidating = false
+    override fun invalidate(action: Action, world: World, origin: BlockPos) {
+        invalidating = true
+        val operator = Operator(action)
+        val originNode: INetworkNode? = NetworkUtils.getNodeFromBlockEntity(world.getBlockEntity(origin))
+        if (originNode is INetworkNodeVisitor) {
+            (originNode as INetworkNodeVisitor).visit(operator)
+        }
+        var currentVisitor: Visitor
+        while (operator.toCheck.poll().also { currentVisitor = it } != null) {
+            currentVisitor.visit(operator)
+        }
```

Comment:
Doesn't really look like the "kotlin way of doing things".
It's weird that kotlin does not provide this in the stdlib
I would suggest
```kotlin
while (true) {
    val currentVisitor = operator.toCheck.poll()?:break
    currentVisitor.visit(operator)
}
```
or even
```kotlin
while (true) {
    operator.toCheck.poll()?.visit(operator)?:break
}
```
or if you want a 1 liner ... with a suppress statement
```kotlin
@Suppress("ControlFlowWithEmptyBody")
while (operator.toCheck.poll()?.visit(operator) != null);
```
although I just realised that this is an ArrayDeque so not concurrent so this might be the best:
```kotlin
val toCheck = operator.toCheck
while (!toCheck.isEmpty()) {
    operator.toCheck.remove().visit(operator)
}
```